### PR TITLE
[BE]: Improve decorator typing

### PR DESCRIPTION
**Pull request by Skylion007**
This fixes decorator typing which unmasks a lot of typing issues in the codebase

cc @ezyang @malfet @xuzhao9 @gramster

### CODE REVIEW

File: torch/_jit_internal.py [hunk 1]
```diff
@@ -707,11 +712,11 @@ def unused_method(self, x):
         # any compiled methods and wasn't decorated with `@torch.jit.export`
         m = torch.jit.script(MyModule())
     """
-    fn._torchscript_modifier = FunctionModifiers.EXPORT
+    fn._torchscript_modifier = FunctionModifiers.EXPORT  # type:ignore[attr-defined]
```
Comment (rec) [inline torch/_jit_internal.py:715]: All right, I see how this movie is going to play out and unfortunately, I think we're going to see this in a lot of the codebase as we move forward, because attaching an attribute to some function that didn't have it before is pretty endemic in this codebase. 😁 

And in the context of what we're doing, it's the right thing to do. My belief is that almost all the time we do this, it's because we've just wrapped a function or something that creates a brand new function and we need to give it "metadata".

So how to fix this problem, with an eye to the general case?

What about a `Protocol`? Is there a way to have a protocol that's callable with just one member, and a `TypeGuard`?

That was the best I came up with unfortunately.

**Review by rec — COMMENTED** [review]
I also think you should change the message to just be "Improve decorator typing..." because it definitely improves the typing!

File: torch/_jit_internal.py (hunk 1 above)
Comment (Skylion007) [reply torch/_jit_internal.py:715]: Something like:
```
class WithMeta(Protocol[P, R]):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R: ...
    __meta__: dict[str, Any]  # or whatever shape your metadata has
```

Comment (Skylion007) [reply torch/_jit_internal.py:715]: @rec Feel free to this cleanup in a follow up PR, this is unblocks a lot of type hints in the codebase so I'd like to merge this first before we optimize it to use Protocols

File: torch/nn/utils/parametrize.py [hunk 2]
```diff
@@ -697,6 +697,7 @@ def remove_parametrizations(
                 if type(original) is torch.Tensor:
                     _maybe_set(original, t)
                 else:
+                    assert isinstance(original, torch.Tensor)
```
Comment (albanD) [inline torch/nn/utils/parametrize.py:700]: Can we more this after the getattr above and raise an error saying that the parametrization didn't return a Tensor as the error message?

File: torch/ao/quantization/fx/lstm_utils.py [hunk 3]
```diff
@@ -104,7 +104,8 @@ def make_qconfig(obs_ctr: _PartialWrapper) -> QConfig:
     # Insert observers into each LSTM cell
     # TODO: maybe make this work for layer_bw as well
     for layer in quantizable_lstm.layers:
-        cell = layer.layer_fw.cell
+        cell = layer.layer_fw.cell  # type: ignore[union-attr]
+        assert isinstance(cell, torch.nn.Module)
```
Comment (albanD) [inline torch/ao/quantization/fx/lstm_utils.py:108]: Please add nice error message here as well saying that cell should be a Module, did you monkey patch it?

**Review by albanD — APPROVED** [review]
Sounds good, please add nice error message for asserts!

File: torch/_jit_internal.py (hunk 1 above)
Comment (rec) [reply torch/_jit_internal.py:715]: Absolutely. I have a tendency to try to put too much into one pot when cooking. 😁 

Thanks for doing this, and for changing the message!

File: torch/nn/utils/parametrize.py (hunk 2 above)
Comment (Skylion007) [reply torch/nn/utils/parametrize.py:700]: We already do a type is check?

**Comment by Skylion007** [comment]
That indicates that there previously type errors internally that were masked by these decorators. @albanD can you find someone to merge an fix these?

**Comment by Skylion007** [comment]
@aorenste Can you help shepherd this in?

**Comment by aorenste** [comment]
> @aorenste Can you help shepherd this in?

Yes - I'll take a look

**Comment by Skylion007** [comment]
Superceded by #154515
